src/fcmp_pp/ffi_types.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2024, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "fcmp_pp_rust/fcmp++.h"
+#include "misc_log_ex.h"
+
+namespace fcmp_pp
+{
+namespace ffi
+{
+//----------------------------------------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------------------------------------
+
+// FFI types instantiated on the Rust side must be free'd back on the Rust side. We wrap them in a unique ptr with a
+// custom deleter that calls the respective Rust free fn.
+#define DEFINE_FCMP_FFI_TYPE(raw_t, new_fn, free_fn)                                              \
```

Comment:
Personally, I might drop the `new_fn` since for some types it doesn't really make sense to have a default value. What is the "default tree root"? What is the default outputs blind?

Reply:
This is also why I argued earlier for the Rust FFI functions returning a `NULL` pointer on an error, instead of returning an int and setting the pointer by reference. Makes passing it to a `std::unique_ptr` more ergonomic. We could use this [errno crate](https://docs.rs/errno/latest/errno/) for setting `errno`, just like a real C stdlib function, if passing error codes is useful for recovering for that specific function.

Reply:
I think for other functions, like `path_new` for example (or whatever it ends up being called), we can definitely expect multiple error cases (failed point decompression, output idx too high). I don't love the idea of adding another dependency to handle something that could be handled pretty simply in imo a clear functional way. And so considering that, we'd expect a separate "constructor" fn approach somewhere. And as long as we do expect it, why not generalize the approach in this way to smoothly support more error cases for functions like this in the future?

I hear you on there not being a default in those cases, but does that matter? I'll just do as you propose if you feel strongly on it.